File: content/problems/0816-ambiguous-coordinates/solutions/split-and-valid-number-forms/Solution.java
class Solution {
    public List<String> ambiguousCoordinates(String s){
        String digits=s.substring(1,s.length()-1);
        List<String> answer=new ArrayList<>();
        for(int split=1;split<digits.length();split++)for(String left:forms(digits.substring(0,split)))for(String right:forms(digits.substring(split)))answer.add("("+left+", "+right+")");
        return answer;
    }
    private List<String> forms(String part){
        List<String> result=new ArrayList<>();
        if(part.length()==1||part.charAt(0)!='0')result.add(part);
        for(int dot=1;dot<part.length();dot++){
            String integer=part.substring(0,dot),fraction=part.substring(dot);
            if((integer.length()==1||integer.charAt(0)!='0')&&fraction.charAt(fraction.length()-1)!='0')result.add(integer+"."+fraction);
        }
        return result;
    }
}

File: content/problems/0816-ambiguous-coordinates/solutions/split-and-valid-number-forms/Solution.kt
class Solution {
    fun ambiguousCoordinates(s:String):List<String>{
        val digits=s.substring(1,s.length-1);
        val answer=mutableListOf<String>();
        for(split in 1 until digits.length)for(left in forms(digits.substring(0,split)))for(right in forms(digits.substring(split)))answer.add("($left, $right)");
        return answer
    };
    private fun forms(part:String):List<String>{
        val result=mutableListOf<String>();
        if(part.length==1||part[0]!='0')result.add(part);
        for(dot in 1 until part.length){
            val integer=part.substring(0,dot);
            val fraction=part.substring(dot);
            if((integer.length==1||integer[0]!='0')&&fraction.last()!='0')result.add("$integer.$fraction")
        };
        return result
    }
}
